dart/collision/CollisionFilter.cpp

```diff
@@ -38,7 +38,125 @@ namespace dart {
+void CompositeCollisionFilter::addCollisionFilter(CollisionFilter* filter)
+{
+  if (!filter)
+    return;
+
+  const auto result = std::find(mFilters.begin(), mFilters.end(), filter);
+  const bool found = (result != mFilters.end());
+  if (found)
+    return;
+
+  mFilters.push_back(filter);
+}
+
+//==============================================================================
+void CompositeCollisionFilter::removeCollisionFilter(
+    const CollisionFilter* filter)
+{
+  mFilters.erase(
+        std::remove(mFilters.begin(), mFilters.end(), filter), mFilters.end());
+}
+
+//==============================================================================
+void CompositeCollisionFilter::removeAllCollisionFilters()
+{
+  mFilters.clear();
+}
+
+//==============================================================================
+bool CompositeCollisionFilter::needsCollisionCheck(
+    const CollisionObject* object1, const CollisionObject* object2) const
+{
+  for (const auto* filter : mFilters)
+  {
+    const bool needCollision = filter->needsCollisionCheck(object1, object2);
+    if (!needCollision)
+      return false;
+  }
+
+  return true;
+}
+
+//==============================================================================
+void BodyNodeCollisionFilter::addBodyNodePairToBlackList(
+    const dynamics::BodyNode* bodyNode1, const dynamics::BodyNode* bodyNode2)
+{
+  if (!bodyNode1 || !bodyNode2)
+    return;
+
+  const auto* bodyNodeLess = bodyNode1;
+  const auto* bodyNodeGreater = bodyNode2;
+
+  if (bodyNodeLess > bodyNodeGreater)
+    std::swap(bodyNodeLess, bodyNodeGreater);
+
+  // Add the pair only when it doesn't already exist
+  const auto resultLeft = mBlackList.find(bodyNodeLess);
+  const bool foundLeft = (resultLeft != mBlackList.end());
+  if (!foundLeft)
+  {
+    mBlackList[bodyNodeLess] = {bodyNodeGreater};
+  }
+  else
+  {
+    auto& key = resultLeft->second;
+
+    const auto resultRight = key.find(bodyNodeGreater);
+    const bool foundRight = (resultRight != key.end());
+    if (!foundRight)
+      key.insert(bodyNodeGreater);
+  }
+}
+
+//==============================================================================
+void BodyNodeCollisionFilter::removeBodyNodePairFromBlackList(
+    const dynamics::BodyNode* bodyNode1, const dynamics::BodyNode* bodyNode2)
+{
+  if (!bodyNode1 || !bodyNode2)
+    return;
+
+  const auto* bodyNodeLess = bodyNode1;
+  const auto* bodyNodeGreater = bodyNode2;
+
+  if (bodyNodeLess > bodyNodeGreater)
+    std::swap(bodyNodeLess, bodyNodeGreater);
+
+  // Remove the pair only when it already exists
+  const auto resultLeft = mBlackList.find(bodyNodeLess);
+  const bool foundLeft = (resultLeft != mBlackList.end());
+  if (foundLeft)
+  {
+    auto& key = resultLeft->second;
+
+    const auto resultRight = key.find(bodyNodeGreater);
+    const bool foundRight = (resultRight != key.end());
+    if (foundRight)
+    {
+      key.erase(bodyNodeGreater);
```

Comment:
One more nitpick: We don't need to check for existence before calling `erase(~)`. If the value doesn't exist in the set, it won't erase anything. We can get rid of the if-statement and just call `key.erase(bodyNodeGreater);` and then `if(key.empty()) mBlackList.erase(resultLeft);`.